docs: upstream nanotee's Lua guide (#21137)
neovim/neovim: runtime/doc/lua-guide.txt

*lua-guide.txt*	Nvim

			    NVIM REFERENCE MANUAL

			  Guide to using Lua in Nvim

This chapter will give you the basics of using Lua in Neovim.

				       Type |gO| to see the table of contents.

==============================================================================
Introduction							     *lua-guide*

COMMON QUESTIONS

Maybe this section isn't needed. We'll see.

- Is vimscript being deprecated?

Nope. Vimscript is being promoted. Vim9script however is deprecated.

- Is my configuration gonna be faster?

Probably not.

- Explain difference between api vs lua api.
 api for long-term guarantees it'll work, lua api for convenience. Plugins
 should probably use api and configs should probably use lua api

Probably way more, blanking out at the moment.

==============================================================================
Using Lua from Neovim

> You can use `:lua` to run lua code. >

Reviewer: Lua is a proper noun and should be capitalized anywhere it's used in prose (this is inconsistent throughout the docs, but for new docs we can be good little grammarians 😊).

```suggestion
You can use `:lua` to run Lua code. >
```

 should probably use api and configs should probably use lua api

Probably way more, blanking out at the moment.

==============================================================================
Using Lua from Neovim

You can use `:lua` to run Lua code. >

	:lua print("Hello!")

You can also use `:lua=`, which is same as `:lua print(vim.inspect(expr))`, to
conveniently check the value of a variable or a table. >

	:lua= package

- Mention heredocs?
- putting Lua files in rtp
- :source vs. :luafile vs. require

==============================================================================
Calling Vimscript from Lua
- vim.cmd
- vim.fn
- vim.api

==============================================================================
Setting variables
- vim.g
- vim.b
- vim.w
- vim.t
- vim.v
- vim.env

==============================================================================
Thermonuclear disaster/Setting options
- vim.opt
- vim.opt_global
- vim.opt_local
	- get()
	- append()
	- prepend()
	- remove()
- vim.o
- vim.bo
- vim.wo
- vim.go

==============================================================================
Defining mappings

> TODO: not a fan of both the API and lua, yeet the API

Reviewer 2: Yes, enough to cover `vim.keymap` here.

API FUNCTIONS

Neovim provides a list of API functions to set, get and delete mappings:

- Global mappings:

>     - `vim.api.nvim_set_keymap()`

Reviewer 3: Tags here maybe?
```suggestion
    - vim.api.|nvim_set_keymap()|
```

    - `vim.api.nvim_get_keymap()`
    - `vim.api.nvim_del_keymap()`
- Buffer-local mappings:
    - `vim.api.nvim_buf_set_keymap()`
    - `vim.api.nvim_buf_get_keymap()`
    - `vim.api.nvim_buf_del_keymap()`

Let's start with `vim.api.nvim_set_keymap()` and
`vim.api.nvim_buf_set_keymap()`

The first argument passed to the function is a string containing the name of
the mode for which the mapping will take effect:

| String value           | Help page     | Affected modes                           | Vimscript equivalent |
| ---------------------- | ------------- | ---------------------------------------- | -------------------- |

> | `''` (an empty string) | `mapmode-nvo` | Normal, Visual, Select, Operator-pending | `:map`               |

Reviewer 3: Double quotes instead of single quotes, `||` for tags
```suggestion
| "" (an empty string)   | |mapmode-nvo| | Normal, Visual, Select, Operator-pending | |:map|               |
```

| "" (empty string)      | |mapmode-nvo|   | Normal, Visual, Select, Operator-pending | |:map|                  |
| `"n"`                    | |mapmode-n|     | Normal                                   | |:nmap|                 |
| `"v"`                    | |mapmode-v|     | Visual and Select                        | |:vmap|                 |
| `"s"`                    | |mapmode-s|     | Select                                   | |:smap|                 |
| `"x"`                    | |mapmode-x|     | Visual                                   | |:xmap|                 |
| `"o"`                    | |mapmode-o|     | Operator-pending                         | |:omap|                 |
| `"!"`                    | |mapmode-ic|    | Insert and Command-line                  | |:map!|                 |
| `"i"`                    | |mapmode-i|     | Insert                                   | |:imap|                 |
| `"l"`                    | |mapmode-l|     | Insert, Command-line, Lang-Arg           | |:lmap|                 |
| `"c"`                    | |mapmode-c|     | Command-line                             | |:cmap|                 |

> | `"t"`                    | |mapmode-t|     | Terminal                                 | |:tmap|                 |

Reviewer 4: The non-empty strings have backticks but the empty string does not. I'd suggest removing these backticks.

Reviewer 2: Do we want to duplicate Vim manual information? I'd say focus on the Lua specifics and link to the corresponding Vim help for basic information. (I don't think the goal is to give an introduction to Vim _and_ Neovim's Lua at the same time.)

Author: Yeah, was thinking this table is a bit much but decided last minute to not remove it. Will a simple

"The prefixes are the ones listed in |:map-modes|, or "!" for |:map!|, or empty fstring for |:map|."

suffice?

The second argument is the left-hand side of the mapping. An empty string is
equivalent to `<Nop>`, which disables a key.

The third argument is the right-hand side of the mapping, which can either be
a string or a Lua function: >lua

	vim.keymap.set('n', '<Leader>ex1', '<Cmd>echomsg "Example 1"<CR>')
	vim.keymap.set('n', '<Leader>ex2', function() print("Example 2") end)
	vim.keymap.set('n', '<Leader>pl1', require('plugin').plugin_action)
	-- To avoid the startup cost of requiring the module, you can wrap it in a function to require it lazily when invoking the mapping:
	vim.keymap.set('n', '<Leader>pl2', function() require('plugin').plugin_action() end)

The fourth (optional) argument is a table containing boolean options for the
mapping as defined in |:map-arguments|, and also `replace_keycodes`,

> `noremap`, `remap` and `desc`. >lua

Reviewer 4: ```suggestion
`remap` and `desc`. >lua
```

Reviewer 2: (`vim.keymap.set` only uses `remap` (which is false by default, hence behaves as `Xnoremap`; to get `Xmap`, you need to pass `{remap = true}`)

| `"t"`                    | |mapmode-t|     | Terminal                                 | |:tmap|                 |

The second argument is the left-hand side of the mapping. An empty string is
equivalent to `<Nop>`, which disables a key.

The third argument is the right-hand side of the mapping, which can either be
a string or a Lua function: >lua

	vim.keymap.set('n', '<Leader>ex1', '<Cmd>echomsg "Example 1"<CR>')
	vim.keymap.set('n', '<Leader>ex2', function() print("Example 2") end)
	vim.keymap.set('n', '<Leader>pl1', require('plugin').plugin_action)
	-- To avoid the startup cost of requiring the module, you can wrap it in a function to require it lazily when invoking the mapping:
	vim.keymap.set('n', '<Leader>pl2', function() require('plugin').plugin_action() end)

The fourth (optional) argument is a table containing boolean options for the
mapping as defined in |:map-arguments|, and also `replace_keycodes`,
`noremap`, `remap` and `desc`. >lua

	vim.keymap.set('n', '<Leader>ex1', '<Cmd>echomsg "Example 1"<CR>', {buffer = true})
	vim.keymap.set('n', '<Leader>ex2', function() print('Example 2') end, {desc = 'Prints "Example 2" to the message area'})

> Defining keymaps with a Lua function is different from using a string. The

Reviewer 4: ```suggestion
Defining mappings with a Lua function is different from using a string. The
```

usual way to show information about a keymap like `:nmap <Leader>ex1` will not
output useful information (the string itself), but instead only show "Lua
function". It is recommended to add a `desc` key to describe the behavior of

> your keymap. This is especially important for documenting plugin mappings so

Reviewer 4: ```suggestion
your mapping. This is especially important for documenting plugin mappings so
```

> users can understand the usage of the keymap more easily.

Reviewer 4: ```suggestion
users can understand the usage of the mapping more easily.
```

*lua-guide.txt*	Nvim

			    NVIM REFERENCE MANUAL

			  Guide to using Lua in Nvim

This chapter will give you the basics of using Lua in Neovim.

				       Type |gO| to see the table of contents.

==============================================================================
Introduction							     *lua-guide*

------------------------------------------------------------------------------
Api vs Lua api

Before diving into how to use Lua in Neovim, it's important to first
understand the distinction between the different api's as they drive a lot of
the design decisions.

> Neovim has two api's, we call these the "api" and the "lua api". You may think

Reviewer 3: ```suggestion
Neovim has two api's, we call these the "nvim api" and the "lua api". You may think
```

> Neovim has two api's, we call these the "api" and the "lua api". You may think

Reviewer 2: ```suggestion
Neovim has two APIs, we call these the "Nvim API" and the "Lua API". You may think
```

The |api-contract| is just a formal way of saying "these functions are stable
and expected to work for a very long time". Plugin authors and users can
expect anything they write with these functions to work for a long time.

This stability and rigidity comes at a cost however. Not only does it make it
near impossible to improve previous design decisions as it would break the
contract, it may also lead to overly abstract functions for the more common
use cases.

The solution for this is the Lua api. It has no contract, which means that we
are free to improve and fix previous design blunders. These functions are
typically higher-level functions with simpler syntax and saner defaults. Often
times these are simply wrappers over the api functions meant for casual usage.

By combining these two approaches, we are able to provide the stability plugin
authors require with the api as well as provide the improvements and the
simplicity the casual user wishes for with the Lua api. There is of course
nothing wrong with plugin authors using the Lua api nor casual users using the
api, the important part is that they're aware of the tradeoff.

> This guide will primarily focus on using the Lua api functions if possible, as
> they tend to be more convenient and more accessible. Not all api functions
> have a Lua api equivalent wrapper. This means you'll need to use the api for
> these cases.

Reviewer 5: ```suggestion
This guide will primarily focus on using the Lua API functions if possible, as
they tend to be more convenient and more accessible. Not all API functions
have a Lua API equivalent wrapper. This means you'll need to use the API for
these cases.
```

==============================================================================
Using Lua from Neovim

You can use `:lua` to run Lua code. >

	:lua print("Hello!")

You can also use `:lua=`, which is same as `:lua print(vim.inspect(expr))`, to
conveniently check the value of a variable or a table. >

	:lua= package

- Mention heredocs?
- putting Lua files in rtp
- :source vs. :luafile vs. require

==============================================================================

> Calling Vimscript from Lua

Reviewer 5: ```suggestion
Calling Vim script from Lua
```

Bram recently standardized the various spellings to this version.

- vim.cmd
- vim.fn
- vim.api

==============================================================================
Setting variables

TODO: yeet API

------------------------------------------------------------------------------
API

With the exception of predefined Vim variables, they can also be deleted (the

> `:unlet` command is the equivalent in Vimscript). Local variables (`l:`),

Reviewer 5: ```suggestion
`:unlet` command is the equivalent in Vim script). Local variables (`l:`),
```

> `:unlet` command is the equivalent in Vimscript). Local variables (`l:`),
> script variables (`s:`) and function arguments (`a:`) cannot be manipulated as

Reviewer 3: ```suggestion
|:unlet| command is the equivalent in Vimscript). Local variables (|l:|),
script variables (|s:|) and function arguments (|a:|) cannot be manipulated as
```

	vim.api.nvim_buf_del_var(3, 'some_tabpage_variable')

------------------------------------------------------------------------------
### Lua

The most convenient way to set vim variables is to use the `vim.*` wrappers:

- |vim.g|   - global variables (|g:|)
- |vim.b|   - buffer variables (|b:|)
- |vim.w|   - window variables (|w:|)
- |vim.t|   - tabpage variables (|t:|)
- |vim.v|   - predefined Vim variables (|v:|)
- |vim.env| - environment variables >

	vim.g.some_global_variable = {
	    key1 = 'value',
	    key2 = 300
	}

	print(vim.inspect(vim.g.some_global_variable)) -- { key1 = "value", key2 = 300 }

	-- target a specific buffer/window/tabpage
	vim.b[2].myvar = 1

Some variable names may contain characters that cannot be used for identifiers
in Lua. You can still manipulate these variables by using this syntax:
`vim.g['my#variable']`.

To delete one of these variables, simply assign `nil` to it: >

	vim.g.some_global_variable = nil

See also:
- |lua-vim-variables|

------------------------------------------------------------------------------
Caveats

You cannot add/update/delete keys from a dictionary stored in one of these

> variables. For example, this snippet of Vimscript code does not work as

Reviewer 5: ```suggestion
variables. For example, this snippet of Vim script code does not work as
```

expected: >

	let g:variable = {}
	lua vim.g.variable.key = 'a'
	echo g:variable
	" {}

You can use a temporary variable as a workaround: >

	let g:variable = {}
	lua << EOF
	local tmp = vim.g.variable
	tmp.key = 'a'
	vim.g.variable = tmp
	EOF
	echo g:variable
	" {'key': 'a'}

This is a known issue:

- https://github.com/neovim/neovim/issues/12544

==============================================================================
Setting options

> The simplest way to set options is to use the `vim.*o` and `vim.opt*` Lua api
> wrappers. These are wrappers for the api functions describe in the

Reviewer 5: ```suggestion
The simplest way to set options is to use the `vim.*o` and `vim.opt*` Lua API
wrappers. These are wrappers for the API functions described in 
```

|api-options|.

------------------------------------------------------------------------------
vim.*o

These wrappers allow you to manipulate options as if they were variables:

> - |vim.o|  - behaves like `:set`
> - |vim.go| - behaves like `:setglobal`
> - |vim.bo| - behaves like `:setlocal` for buffer-local options
> - |vim.wo| - behaves like `:setlocal` for window-local options >

Reviewer 3: ```suggestion
- |vim.o|  - behaves like |:set|
- |vim.go| - behaves like |:setglobal|
- |vim.bo| - behaves like |:setlocal| for buffer-local options
- |vim.wo| - behaves like |:setlocal| for window-local options >
```

	vim.o.smarttab = false -- :set nosmarttab
	print(vim.o.smarttab) -- false
	vim.o.isfname = vim.o.isfname .. ',@-@' -- on Linux: :set isfname+=@-@
	print(vim.o.isfname) -- '@,48-57,/,.,-,_,+,,,#,$,%,~,=,@-@'

	vim.bo.shiftwidth = 4
	print(vim.bo.shiftwidth) -- 4

You can specify a number for buffer-local and window-local options. If no
number is given, the current buffer/window is used: >

	vim.bo[4].expandtab = true -- sets expandtab to true in buffer 4
	vim.wo.number = true -- sets number to true in current window

------------------------------------------------------------------------------
vim.opt*

> These wrappers are a more programmatic way to set options which is more suited
> for plugin authors.

Reviewer 5: ```suggestion
These wrappers are a more programmatic way to set options which are more suited
for plugin authors.
```

- |vim.opt|: behaves like |:set|
- |vim.opt_global|: behaves like |:setglobal|
- |vim.opt_local|: behaves like |:setlocal| >

	vim.opt.smarttab = false
	print(vim.opt.smarttab:get()) -- false

Some options can be set using Lua tables: >

	vim.opt.completeopt = {'menuone', 'noselect'}
	print(vim.inspect(vim.opt.completeopt:get())) -- { "menuone", "noselect" }

Wrappers for list-like, map-like and set-like options also come with methods
and metamethods that work similarly to their |:set+=|, |:set^=| and |:set-=|

> counterparts in Vimscript. >

Reviewer 5: ```suggestion
counterparts in Vim script. >
```

	vim.opt.shortmess:append({ I = true })
	-- alternative form:
	vim.opt.shortmess = vim.opt.shortmess + { I = true }

	vim.opt.whichwrap:remove({ 'b', 's' })
	-- alternative form:
	vim.opt.whichwrap = vim.opt.whichwrap - { 'b', 's' }

Be sure to look at |vim.opt| for more information.

==============================================================================
Defining mappings

Neovim provides two functions to set and delete mappings, |vim.keymap.set()|
and |vim.keymap.del()|.

vim.keymap.set()
-------------------------------------------------------------------------------

> |vim.keymap.set()| is a convenience function of api functions

Reviewer 5: ```suggestion
|vim.keymap.set()| is a convenience function of API functions
```

vim.api.|nvim_set_keymap()| and vim.api.|nvim_buf_set_keymap()|

The first argument is a string or a table of strings containing the mode
prefix for which the mapping will take effect: >

	-- Normal mode
	vim.keymap.set('n', '<Leader>ex1', '<Cmd>lua vim.notify("Example 1")<CR>')
	-- Normal and Command-line mode
	vim.keymap.set({'n', 'c'}, '<Leader>ex2', '<Cmd>lua vim.notify("Example 2")<CR>')

TODO: this table is a bit much. Maybe just refer to :map-arguments instead and
add that "!" is ":map!" and "" is ":map".

> | String value           | Help page     | Affected modes                           | Vimscript equivalent |

Reviewer 5: ```suggestion
| String value           | Help page     | Affected modes                           | Vim script equivalent |
```

| "" (empty string)      | |mapmode-nvo|   | Normal, Visual, Select, Operator-pending | |:map|                 |
| `'n'`                    | |mapmode-n|     | Normal                                   | |:nmap|                |
| `'v'`                    | |mapmode-v|     | Visual and Select                        | |:vmap|                |
| `'s'`                    | |mapmode-s|     | Select                                   | |:smap|                |
| `'x'`                    | |mapmode-x|     | Visual                                   | |:xmap|                |
| `'o'`                    | |mapmode-o|     | Operator-pending                         | |:omap|                |
| `'!'`                    | |mapmode-ic|    | Insert and Command-line                  | |:map!|                |
| `'i'`                    | |mapmode-i|     | Insert                                   | |:imap|                |
| `'l'`                    | |mapmode-l|     | Insert, Command-line, Lang-Arg           | |:lmap|                |
| `'c'`                    | |mapmode-c|     | Command-line                             | |:cmap|                |

> | `'t'`                    | |mapmode-t|     | Terminal                                 | |:tmap|                |

Reviewer 3: Use double quotes for the rest of string values too

Defining keymaps with a Lua function is different from using a string. The
usual way to show information about a keymap like `:nmap <Leader>ex1` will not
output useful information (the string itself), but instead only show "Lua
function". It is recommended to add a `desc` key to describe the behavior of
your keymap. This is especially important for documenting plugin mappings so
users can understand the usage of the keymap more easily.

Mappings are `noremap` by default: >

    vim.keymap.set('n', '<Leader>test1', '<Cmd>echo "test"<CR>')
    -- :nnoremap <Leader>test <Cmd>echo "test"<CR>

    -- If you DO want the mapping to be recursive, set the "remap" option to "true"
    vim.keymap.set('n', '>', ']', {remap = true})
    -- :nmap > ]

TODO: following seems niche. Maybe remove.

In `expr` mappings, `nvim_replace_termcodes()` is automatically applied to
strings returned from Lua functions: >

    vim.keymap.set('i', '<Tab>', function()
        return vim.fn.pumvisible == 1 and '<C-N>' or '<Tab>'
    end, {expr = true})

vim.keymap.del()
-------------------------------------------------------------------------------

> |vim.keymap.del()| is a convenience function of api functions

Reviewer 5: ```suggestion
|vim.keymap.del()| is a convenience function of API functions
```

	echo g:variable
	" {}

You can use a temporary variable as a workaround: >lua

	let g:variable = {}
	lua << EOF
	local tmp = vim.g.variable
	tmp.key = 'a'
	vim.g.variable = tmp
	EOF
	echo g:variable
	" {'key': 'a'}

This is a known issue:

- https://github.com/neovim/neovim/issues/12544

==============================================================================
Setting options

The simplest way to set options is to use the `vim.*o` and `vim.opt*` Lua API
wrappers. These are wrappers for the API functions described in |api-options|.

------------------------------------------------------------------------------
vim.*o

These wrappers allow you to manipulate options as if they were variables:

- |vim.o|  - behaves like |:set|
- |vim.go| - behaves like |:setglobal|
- |vim.bo| - behaves like |:setlocal| for buffer-local options
- |vim.wo| - behaves like |:setlocal| for window-local options >lua

	vim.o.smarttab = false -- :set nosmarttab
	print(vim.o.smarttab) -- false
	vim.o.isfname = vim.o.isfname .. ',@-@' -- on Linux: :set isfname+=@-@
	print(vim.o.isfname) -- '@,48-57,/,.,-,_,+,,,#,$,%,~,=,@-@'

	vim.bo.shiftwidth = 4
	print(vim.bo.shiftwidth) -- 4

> You can specify a number for buffer-local and window-local options. If no

Reviewer 3: ```suggestion
You can specify a buffer or window number for buffer-local and window-local options. If no
```

	function! s:completion_function(ArgLead, CmdLine, CursorPos) abort
	    return join([
		\ 'strawberry',
		\ 'star',
		\ 'stellar',
		\ ], "\n")
	endfunction

	command! -nargs=1 -complete=custom,s:completion_function Test echo <q-args>
	" Typing `:Test st[ae]<Tab>` returns "star" and "stellar"

Passing a Lua function to `complete` makes it behave like `customlist` which
leaves filtering up to the user: >lua

	vim.api.nvim_create_user_command('Test', function() end, {
	    nargs = 1,
	    complete = function(ArgLead, CmdLine, CursorPos)
		return {
		    'strawberry',
		    'star',
		    'stellar',
		}
	    end,
	})

	-- Typing `:Test z<Tab>` returns all the completion results because the list was not filtered

==============================================================================
Further reading:
TODO: write
• vim.loop
• vim.lsp
• vim.treesitter

==============================================================================

TODO: maybe add a "General tips and recommendations" section like nanotee
does? Depends on how long this will get.

This guide is in large part taken from nanotee's Lua guide:
https://github.com/nanotee/nvim-lua-guide

Thank you nanotee :)

> vim:tw=78:ts=8:noet:ft=help:norl:

Reviewer 3: I'd set the indentation to only spaces instead of mixing tabs and spaces
```suggestion
vim:tw=78:sts=4:sw=4:et:ft=help:norl:
```

Reviewer 2: "Neovim" help files tend to have
```vim
 vim:tw=78:ts=8:sw=4:sts=4:et:ft=help:norl:
```

Reviewer 2: And please, no mixing of tabs and spaces!

vim.api.|nvim_del_keymap()| and vim.api.|nvim_buf_del_keymap()|

|vim.keymap.del()| works the same way as |vim.keymap.set()| but deletes
mappings instead: >lua

	vim.keymap.del('n', '<Leader>ex1')
	vim.keymap.del({'n', 'c'}, '<Leader>ex2', {buffer = true})

See also:
• vim.api.|nvim_get_keymap()|     - list all global mapping
• vim.api.|nvim_buf_get_keymap()| - list all mappings for buffer

==============================================================================
Defining autocommands
TODO: write

Looks useful:
• nvim_del_autocmd

Looks decent:
• nvim_clear_autocmds
• nvim_get_autocmds

Looks meh:
• nvim_create_augroup
• nvim_del_augroup_by_id
• nvim_del_augroup_by_name

wat:
• nvim_exec_autocmds

TODO: what's the point of manually executing an autocommand?

-------------------------------------------------------------------------------
Creating autocommands

Use vim.api.|nvim_create_autocmd()| to create autocommands from Lua. Lua
autocommands can execute a command (like regular autocommands), but they can
also call a function. We call these "callback functions" since we don't
immediately execute them, we only tell Neovim to run them some time in the

> future if the autocommand is triggered.

Reviewer 3: I would explain autocommands as something like "running an action on some event", and then say that it can be an ex command or a lua function. Since this is a guide, if a mechanism can be briefly explained I think we should just do that, instead of assuming that the reader knows vim script and explaining in terms of how it relates to it.

Author: Hmm, I'm not a fan, sorry. I think assuming the reader has some previous knowledge of vim (and stating this at the intro) helps the guide be more focused. I'm not completely against the idea.

@clason thoughts?

Reviewer 2: I agree that it should not start from `:Tutor` and assume some basics, but if you can remind people with a single sentence, that is always welcome.

> An autocommand is a command that is executed automatically in response to some
> event, such as a file being read or written or a buffer change. Use
> vim.api.|nvim_create_autocmd()| to create autocommands from Lua. Lua
> autocommands can execute a command (like regular autocommands), but they can
> also call a function. We call these "callback functions" since we don't
> immediately execute them, we only tell Neovim to run them some time in the
> future if the autocommand is triggered.

Reviewer 3: ```suggestion
An autocommand is a vim command or a lua function that is executed
automatically in response to some event, such as a file being read or
written or a buffer change. Use vim.api.|nvim_create_autocmd()| to
create autocommands from Lua.
```

duplicate functions that does the same thing.

This guide will primarily focus on using the Lua API functions if possible, as
they tend to be more convenient and more accessible.

==============================================================================
Where to put your files

Neovim supports using `init.vim` or `init.lua` as the configuration file, but
not both at the same time. This should be placed in your |'runtimepath'|,
which is typically `~/.config/nvim` for unix-y systems and
`~/AppData/Local/nvim/` for Windows.

If you'd like to run any Lua or Vim script on startup automatically, then you
can simply put them in `plugin/` in your |'runtimepath'|.

See also:
• |config|

------------------------------------------------------------------------------
Lua modules

Lua modules are basically standalone Lua scripts you can import. They should
be located in the `lua/` directory in you |'runtimepath'|. Let's assume you
have the following directory structure:

~/.config/nvim
|-- after/
|-- ftplugin/
|-- lua/
|  |-- myluamodule.lua
|  |-- other_modules/
|     |-- anothermodule.lua
|     |-- init.lua
|-- plugin/
|-- syntax/
|-- init.vim

> 

Reviewer 2: `text` is not a Lua parser -- if no language annotation is specified, it defaults to the usual un-highlighted code blocks.

The following Lua code will load myluamodule.lua: >lua

>     require("myluamodule")

Reviewer 2: Not sure I want to do this here. The style guide is opinionated and can be confusing to newcomers (especially regarding omitting parens). In any case, it only applies to Neovim code, not user config. So as long as it's valid Lua, I don't think we need to care here.

in Lua. You can still manipulate these variables by using this syntax:
`vim.g['my#variable']`.

To delete one of these variables, simply assign `nil` to it: >lua

    vim.g.some_global_variable = nil

See also:
- |lua-vim-variables|

==============================================================================
Setting options

The simplest way to set options is to use the `vim.*o` and `vim.opt*` Lua API
wrappers. These are wrappers for the API functions described in |api-options|.

------------------------------------------------------------------------------
vim.*o

These wrappers allow you to manipulate options as if they were variables:

• |vim.o|  - behaves like |:set|
• |vim.go| - behaves like |:setglobal|
• |vim.bo| - behaves like |:setlocal| for buffer-local options
• |vim.wo| - behaves like |:set| for window-local options >lua

    vim.o.smarttab = false -- :set nosmarttab
    print(vim.o.smarttab) -- false
    vim.o.isfname = vim.o.isfname .. ',@-@' -- on Linux: :set isfname+=@-@
    print(vim.o.isfname) -- '@,48-57,/,.,-,_,+,,,#,$,%,~,=,@-@'

    vim.bo.shiftwidth = 4
    print(vim.bo.shiftwidth) -- 4

> You can specify a buffer or window number for buffer-local and window-local

Reviewer 4: `vim.wo` can only be indexed using window ID, not window number.

nothing wrong with casual users/non-plugin authors using API functions. Quite
the opposite: not all API functions have a Lua API equivalent. Either they
haven't been created yet or they simply aren't needed. A Lua API function is
only needed if it significantly simplifies using the API functions in some
aspect. Having a wrapper for the sake of it would only result in a bunch of
duplicate functions that does the same thing.

This guide will primarily focus on using the Lua API functions if possible, as
they tend to be more convenient and more accessible.

==============================================================================
Where to put your files

Neovim supports using `init.vim` or `init.lua` as the configuration file, but
not both at the same time. This should be placed in your |'runtimepath'|,
which is typically `~/.config/nvim` for unix-y systems and
`~/AppData/Local/nvim/` for Windows.

If you'd like to run any Lua or Vim script on startup automatically, then you
can simply put them in one of the following directories in your
|'runtimepath'|: `colors/`, `compiler/`, `ftplugin/`, `ftdetect/`, `indent/`,

> `plugin/` and `syntax/`. If you're unsure where to put your script, then it

Reviewer 4: This looks a bit strange.

Author: How would you like it to look?

> should probably be in `plugin/`.

Author: Yeah, that looks better. Thanks.

Take the following example, where both commands does the same thing: >lua

    vim.cmd("%s/\\Vfoo/bar/g")
    vim.cmd([[%s/\Vfoo/bar/g]])

Notice that in the first case we need an extra `\` to escape the backslash.

Another benefit of using double bracketed strings is that they support
multiline string. The following command is perfectly valid: >lua

    vim.cmd([[
        set ignorecase
        colorscheme tokyonight
    ]])

This is analogous to using Lua heredocs. It is entirely possible (except in
rare cases) to surround the contents of your `init.vim` with `vim.cmd([[]])`,
rename `init.vim` to `init.lua` and still have a working configuration. If you
wish to rewrite your configuration in Lua then this approach may be less
chaotic than a full rewrite from scratch as it allows you to incrementally
change your configuration.

There is also a third way to use |vim.cmd()|: >lua

    vim.cmd.echo("\"Hello World!\"")

This is essentially syntactic sugar of the non-bracketed string explained at
the start of this section.

------------------------------------------------------------------------------
Calling functions

> Use |vim.fn| to call Vim script commands from Lua. Data types between Lua and

Reviewer 4: ```suggestion
Use |vim.fn| to call Vim script functions from Lua. Data types between Lua and
```

> Vim script are automatically converted: >lua

Reviewer 4: There may be a word missing here.

The second argument is the left-hand side of the mapping. An empty string is
equivalent to |<Nop>|, which disables a key.

The third argument is the right-hand side of the mapping, which can either be
a string or a Lua function: >lua

    vim.keymap.set('n', '<Leader>ex1', '<Cmd>echomsg "Example 1"<CR>')
    vim.keymap.set('n', '<Leader>ex2', function() print("Example 2") end)
    vim.keymap.set('n', '<Leader>pl1', require("plugin").plugin_action)
    -- To avoid the startup cost of requiring the module, you can wrap it in a function to require it lazily when invoking the mapping:
    vim.keymap.set('n', '<Leader>pl2', function() require("plugin").plugin_action() end)

The fourth (optional) argument is a table containing boolean options for the
mapping as defined in |:map-arguments|, and also `replace_keycodes`,
`remap` and `desc`. >lua

    vim.keymap.set('n', '<Leader>ex1', '<Cmd>echomsg "Example 1"<CR>', {buffer = true})
    vim.keymap.set('n', '<Leader>ex2', function() print('Example 2') end, {desc = 'Prints "Example 2" to the message area'})

Defining mappings with a Lua function is different from using a string. The

> usual way to show information about a keymap like `:nmap <Leader>ex1` will not

Reviewer 4: ```suggestion
usual way to show information about a mapping like `:nmap <Leader>ex1` will not
```

should probably be in `plugin/`.

See also:
• |config|

------------------------------------------------------------------------------
Lua modules

Lua modules are basically standalone Lua scripts you can import. They should
be located in the `lua/` directory in you |'runtimepath'|. Let's assume you
have the following directory structure:

~/.config/nvim
|-- after/
|-- ftplugin/
|-- lua/
|  |-- myluamodule.lua
|  |-- other_modules/
|     |-- anothermodule.lua
|     |-- init.lua
|-- plugin/
|-- syntax/
|-- init.vim

The following Lua code will load myluamodule.lua: >lua

    require("myluamodule")

Notice the absence of a .lua extension.

Similarly, loading `other_modules/anothermodule.lua` is done like so: >lua

    require('other_modules.anothermodule')
    -- or
    require('other_modules/anothermodule')

Path separators are denoted by either a dot `.` or a slash `/`.

A folder containing an |init.lua| file can be required directly, without
having to specify the name of the file: >lua

    require('other_modules') -- loads other_modules/init.lua

Requiring a nonexistent module or a module which contains syntax errors aborts

> the currently executing script. pcall() may be used to prevent errors. >lua

Reviewer 2: ```suggestion
the currently executing script. `pcall()` may be used to prevent errors. >lua
```

    local ok, _ = pcall(require, 'module_with_error')
    if not ok then
      -- not loaded
    end

See also:

> • |lua-require|

Reviewer 2: ```suggestion
• |lua-require|
• |luaref-pcall()|
```

• |vim.g|   - global variables (|g:|)
• |vim.b|   - buffer variables (|b:|)
• |vim.w|   - window variables (|w:|)
• |vim.t|   - tabpage variables (|t:|)
• |vim.v|   - predefined Vim variables (|v:|)
• |vim.env| - environment variables >lua

    vim.g.some_global_variable = {
        key1 = "value",
        key2 = 300
    }

    print(vim.inspect(vim.g.some_global_variable)) -- { key1 = "value", key2 = 300 }

    -- target a specific buffer/window/tabpage
    vim.b[2].myvar = 1

Some variable names may contain characters that cannot be used for identifiers
in Lua. You can still manipulate these variables by using this syntax:
`vim.g['my#variable']`.

To delete one of these variables, simply assign `nil` to it: >lua

    vim.g.some_global_variable = nil

See also:
- |lua-vim-variables|

==============================================================================
Setting options

The simplest way to set options is to use the `vim.*o` and `vim.opt*` Lua API
wrappers. These are wrappers for the API functions described in |api-options|.

------------------------------------------------------------------------------
vim.*o

> These wrappers allow you to manipulate options as if they were variables:

Author: Hmm, do you mean draw parallel with `:let &option` as well as `:set`, or replace the `:set` parallel with the `let` parallel?

> These wrappers allow you to manipulate options as if they were variables:

Reviewer 2: No, I mean the following: in vimscript, you can do `:set foo` or `:let &foo`. I consider `vim.opt` and `vim.o` to be the Lua analogue of this, respectively.

(This completely orthogonal to `:set`  vs. `:setlocal`.)

• |vim.o|  - behaves like |:set|
• |vim.go| - behaves like |:setglobal|
• |vim.bo| - behaves like |:setlocal| for buffer-local options

> • |vim.wo| - behaves like |:setlocal| for window-local options >lua

Reviewer 2: ```suggestion
• |vim.wo| - behaves like |:set| for window-local options >lua
```

Reviewer 2: Careful, window-local options are nasty...

Author: I went with my intuition :)

    vim.o.smarttab = false -- :set nosmarttab
    print(vim.o.smarttab) -- false
    vim.o.isfname = vim.o.isfname .. ',@-@' -- on Linux: :set isfname+=@-@
    print(vim.o.isfname) -- '@,48-57,/,.,-,_,+,,,#,$,%,~,=,@-@'

    vim.bo.shiftwidth = 4
    print(vim.bo.shiftwidth) -- 4

You can specify a buffer or window number for buffer-local and window-local
options. If no number is given, the current buffer/window is used: >lua

    vim.bo[4].expandtab = true -- sets expandtab to true in buffer 4
    vim.wo.number = true -- sets number to true in current window
<

------------------------------------------------------------------------------
vim.opt*

These wrappers are a more programmatic way to set options which are more
suited for plugin authors.

> TODO: can use tags after https://github.com/neovim/neovim/pull/21138 is
> merged.
> 
> • |vim.opt|: behaves like |:set|
> • vim.opt_global: behaves like |:setglobal|
> • vim.opt_local: behaves like |:setlocal| >lua

Reviewer 2: ```suggestion
• |vim.opt|: behaves like |:set|
• |vim.opt_global|: behaves like |:setglobal|
• |vim.opt_local|: behaves like |:setlocal| >lua
```
I just merged it :)

(And `opt_local` _does_ behave like `:setlocal` also for window-local options. Maybe it would make sense here to start this section with `vim.opt` and then introduce `vim.*o` as an advanced, more direct, "variable-like" access?

This guide will go through the basics of using Lua in Neovim. It is not meant
to be a comprehensive encyclopedia of all available features, nor will it
detail all intricacies. Think of it as a survival kit - the bare minimum
needed to know to comfortably get started on using Lua in Neovim.

An important thing to note is that this isn't a guide to the Lua language
itself. Rather, this is a guide on how to configure and modify Neovim through
the Lua language and the functions we provide to help with this. Take a look
at |luaref| and |lua-concepts| if you'd like to learn more about Lua itself.
Similarly, this guide assumes some familiarity with the basics of Neovim
(commands, options, mappings, autocommands), which are covered in the
|user-manual|.

------------------------------------------------------------------------------
Some words on the API                                            *lua-guide-api*

The purpose of this guide is to introduce the different ways of interacting
with Neovim through Lua (the "API"). This API consists of three different
layers:

1. The "Vim API" inherited from Vim: |ex-commands| and |builtin| as well as

> |user-function|s in Vimscript. These are accessed through |vim.cmd| and

Reviewer 5: ```suggestion
|user-function|s in Vim script. These are accessed through |vim.cmd| and
```

Official spelling is spaced now, Bram fixed it throughout docs (hopefully everywhere). It hasn't been fixed yet in our Neovim-specific docs.

|vim.fn| respectively, which are discussed under |lua-guide-vimscript| below.

2. The "Neovim API" written in C for use in remote plugins and GUIs; see |api|.
These functions are accessed through |vim.api|.

3. The "Lua API" written in and specifically for Lua. These are any other
functions accessible through `vim.*` not mentioned already.

This distinction is important, as API functions inherit behavior from their
original layer: For example, Neovim API functions always need all arguments to
be specified even if Lua itself allows omitting arguments (which are then
passed as `nil`); and Vim API functions can use 0-based indexing even if Lua
arrays are 1-indexed by default.

Through this, any possible interaction can be done through Lua without writing
a complete new API from scratch. For this reason, functions are usually not
duplicated between layers unless there is a significant benefit in
functionality or performance (e.g., you can map Lua functions directly through
|nvim_create_autocmd| but not through |:autocmd|). In case there are multiple
ways of achieving the same thing, this guide will only cover what is most

> convenient to from Lua.

Reviewer 5: ```suggestion
convenient to use from Lua.
```

==============================================================================
Using Lua                                                  *lua-guide-using-Lua*

To run Lua code from the Neovim command line, use the |:lua| command:
>vim
    :lua print("Hello!")
<
Note: each `:lua` command has its own scope and variables declared with the
local keyword are not accessible outside of the command. This won't work:
>vim
    :lua local foo = 1
    :lua print(foo)
    " prints "nil" instead of "1"
<

> You can also use `:lua =`, which is same as `:lua print(vim.inspect(expr))`, to

Reviewer 5: ```suggestion
You can also use `:lua =`, which is the same as `:lua print(vim.inspect(expr))`, to
```

conveniently check the value of a variable or a table:
>lua
    :lua =package
<
To run a Lua script in an external file, you can use the |:source| command

> exactly like for a Vimscript:

Reviewer 5: ```suggestion
exactly like you do in Vim script:
```

>vim
    :source ~/programs/baz/myluafile.lua
<

> Finally, you can include Lua code in a vimscript file by putting it inside a

Reviewer 5: ```suggestion
Finally, you can include Lua code in a Vim script file by putting it inside a
```

|lua-heredoc| block:
>vim
    lua << EOF
        local tbl = {1, 2, 3}
        for k, v in ipairs(tbl) do
            print(v)
        end
    EOF
<
------------------------------------------------------------------------------
Using Lua files on startup                                    *lua-guide-config*

Neovim supports using `init.vim` or `init.lua` as the configuration file, but
not both at the same time. This should be placed in your |config| directory,
which is typically `~/.config/nvim` for Linux, BSD, or macOS, and
`~/AppData/Local/nvim/` for Windows. Note that you can use Lua in `init.vim`

> and Vimscript in `init.lua`, which will be covered below.

Reviewer 5: ```suggestion
and Vim script in `init.lua`, which will be covered below.
```

If you'd like to run any other Lua script on |startup| automatically, then you

> can simply put it in `plugin/` in your |'runtimepath'|.

Reviewer 5: ```suggestion
can simply put it in a Lua script inside of `plugin/` on your |'runtimepath'|.
```

------------------------------------------------------------------------------
Lua modules                                                  *lua-guide-modules*

If you want to load Lua files on demand, you can place them in the `lua/`

> directory in your |'runtimepath'| and load them with `require`. (This is the
> Lua equivalent of Vimscript's |autoload|.)

Reviewer 5: ```suggestion
directory in your |'runtimepath'| and load them with `require()`. This is similar 
to the Vim script |autoload| mechanism.
```

Unless it's exactly the same, which I don't think it is, I don't think we should say 'equivalent' here,  aren't there lots of situations and small nuances between the two loading mechanisms?

Reviewer 2: The main point (on demand vs. on startup) is indeed exactly the same.

Let's assume you have the following directory structure:
>
    ~/.config/nvim
    |-- after/
    |-- ftplugin/
    |-- lua/
    |  |-- myluamodule.lua
    |  |-- other_modules/
    |     |-- anothermodule.lua
    |     |-- init.lua
    |-- plugin/
    |-- syntax/
    |-- init.vim
<

Then the following Lua code will load `myluamodule.lua`:
>lua
    require("myluamodule")
<
Note the absence of a `.lua` extension.

> Similarly, loading `other_modules/anothermodule.lua` is done via

Reviewer 5: ```suggestion
Similarly, loading `other_modules/anothermodule.lua` is done via:
```

>lua
    require('other_modules/anothermodule')
    -- or
    require('other_modules.anothermodule')
<

Note how "submodules" are just subdirectories; the `.` is equivalent to the
path separator `/` (even on Windows).

A folder containing an |init.lua| file can be required directly, without
having to specify the name of the file:
>lua
    require('other_modules') -- loads other_modules/init.lua
<
Requiring a nonexistent module or a module which contains syntax errors aborts

> the currently executing script. `pcall()` may be used to prevent errors.

Reviewer 5: ```suggestion
the currently executing script. `pcall()` may be used to prevent errors:
```

>lua
    local ok, _ = pcall(require, 'module_with_error')
    if not ok then
      -- not loaded
    end
<
In contrast to |:source|, |require()| not only searches through all `lua/` directories

> under |'runtimepath'|, it also cache the module on first use. Calling

Reviewer 5: ```suggestion
under |'runtimepath'|, it also caches modules on first use. Calling
```

To run an arbitrary Vim command from Lua, pass it as a string to |vim.cmd()|:
>lua
    vim.cmd("colorscheme habamax")
<
Note that special characters will need to be escaped with backslashes:
>lua
    vim.cmd("%s/\\Vfoo/bar/g")
<
An alternative is to use a literal string (see |luaref-literal|) delimited by
double brackets `[[ ]]` as in
>lua
    vim.cmd([[%s/\Vfoo/bar/g]])
<
Another benefit of using literal strings is that they can be multiple lines;
this allows you to pass multiple commands to a single call of |vim.cmd()|:
>lua
    vim.cmd([[
        highlight Error guibg=red
        highlight link Warning Error
    ]])
<
This is the converse of |lua-heredoc| and allows you to include Lua code in
your `init.vim`.

If you want to build your Vim command programmatically, the following form can
be useful (all these are equivalent to the corresponding line above):
>lua
    vim.cmd.colorscheme("habamax")
    vim.cmd.highlight({ "Error", "guibg=red" })
    vim.cmd.highlight({ "link", "Warning", "Error" })
<
------------------------------------------------------------------------------
Vimscript functions                                    *lua-guide-vim-functions*

> Use |vim.fn| to call Vimscript functions from Lua. Data types between Lua and

Reviewer 5: ```suggestion
Use |vim.fn| to call Vim script functions from Lua. Data types between Lua and
```

<
Note that you cannot directly change fields of array variables. This won't
work:
>lua
    vim.g.some_global_variable.key2 = 400
    vim.pretty_print(vim.g.some_global_variable)
    --> { key1 = "value", key2 = 300 }
<
Instead, you need to create an intermediate Lua table and change this:
>lua
    local temp_table = vim.g.some_global_variable
    temp_table = keys = 400
    vim.g.some_global_variable = temp_table
    vim.pretty_print(vim.g.some_global_variable)
    --> { key1 = "value", key2 = 400 }
<
To delete a variable, simply set it to `nil`:
>lua
    vim.g.myvar = nil
<

See also:
• |lua-vim-variables|

==============================================================================
Options                                                      *lua-guide-options*

There are two complementary ways of setting |options| via Lua.

------------------------------------------------------------------------------
vim.opt

The most convenient way for setting global and local options, e.g., in `init.lua`,
is through `vim.opt` and friends:

• |vim.opt|        - behaves like |:set|
• |vim.opt_global| - behaves like |:setglobal|
• |vim.opt_local|  - behaves like |:setlocal|

> For example, the Vimscript commands

Reviewer 5: ```suggestion
For example, the Vim script commands
```

>vim
    set smarttab
    set nosmarttab
<
are equivalent to
>lua
    vim.opt.smarttab = true
    vim.opt.smarttab = false
<
In particular, they allow an easy way to working with list-like, map-like, and

> set-like options through Lua tables: Instead of

Reviewer 5: ```suggestion
set-like options through Lua tables. Instead of:
```

>vim
    set wildignore=*.o,*.a,__pycache__
    set listchars=space:_,tab:>~
    set formatoptions=njt
<

> you can use

Reviewer 5: ```suggestion
you can use:
```

your mapping. This is especially important for documenting plugin mappings so
users can understand the usage of the mapping more easily.

Mappings are `noremap` by default: >lua

    vim.keymap.set('n', '<Leader>test1', '<Cmd>echo "test"<CR>')
    -- :nnoremap <Leader>test <Cmd>echo "test"<CR>

    -- If you DO want the mapping to be recursive, set the "remap" option to "true"
    vim.keymap.set('n', '>', ']', {remap = true})
    -- :nmap > ]

------------------------------------------------------------------------------
vim.keymap.del()

|vim.keymap.del()| is a convenience function of API functions
vim.api.|nvim_del_keymap()| and vim.api.|nvim_buf_del_keymap()|

|vim.keymap.del()| works the same way as |vim.keymap.set()| but deletes
mappings instead: >lua

    vim.keymap.del('n', '<Leader>ex1')
    vim.keymap.del({'n', 'c'}, '<Leader>ex2', {buffer = true})

------------------------------------------------------------------------------
See also
• vim.api.|nvim_get_keymap()|     - list all global mapping
• vim.api.|nvim_buf_get_keymap()| - list all mappings for buffer

==============================================================================
Autocommands

> An autocommand is a vim command or a lua function that is executed

Reviewer 5: ```suggestion
An autocommand is a Vim command or a Lua function that is executed
```

or a buffer change.

------------------------------------------------------------------------------
Creating autocommands

Use vim.api.|nvim_create_autocmd()| to create autocommands from Lua.

The first, and the only required, argument is a string or table of strings of
possible |events|. There are many optional arguments, but the most important
are:

• `pattern`  - string or table of strings with patterns to match against.
             Defaults to `pattern = "*"`. NOTE: Environment variable like
             `$HOME` and `~` aren't automatically expanded, use
             vim.fn.|expand()|.
• `command`  - Vim command, must be a string. Can't be used with `callback`.
• `callback` - Lua function. Can't be used with `command`. >lua

    vim.api.nvim_create_autocmd({"BufEnter", "BufWinEnter"}, {
      pattern = {"*.c", "*.h"},
      command = "echo 'Entering a C or C++ file'",
    })

    -- Same autocommand written with a Lua function instead
    vim.api.nvim_create_autocmd({"BufEnter", "BufWinEnter"}, {
        pattern = {"*.c", "*.h"},
        callback = function() print("Entering a C or C++ file") end,
    })

> Neovim will always call the function on the right with a single table with

Reviewer 5: ```suggestion
Neovim will always call the function on the right with a single table that has
```

        end,
    })

See also:
• |autocmd-groups|

------------------------------------------------------------------------------
Deleting autocommands

You can use vim.api.|nvim_clear_autocmds()| to remove many autocommands at
once based on given option. It accepts only a single optional table with
options >lua

    -- Delete all BufEnter and InsertLeave autocommands
    vim.api.nvim_clear_autocmds({event = {"BufEnter", "InsertLeave"}})

    -- Delete all autocommands that uses "*.py" pattern
    vim.api.nvim_clear_autocmds({pattern = "*.py"})

    -- Delete all autocommands in group "scala"
    vim.api.nvim_clear_autocmds({group = "scala"})

    -- Delete all ColorScheme autocommands in current buffer
    vim.api.nvim_clear_autocmds({event = "ColorScheme", buffer = 0 })

Note: autocommands in groups will only be removed if the `group` key is
specified, even if another option matches it.

------------------------------------------------------------------------------
See also
• |api-autocmd|
• |autocmd|
• |40.3|

==============================================================================
User commands

A user command is a custom Ex command you can define. A user command can act

> just like a built-in command: it can have a range or arguments and arguments

Reviewer 5: ```suggestion
just like a built-in command: it can have a range or arguments, and arguments
```

> can be completed as filenames or buffer names for example.

Reviewer 5: ```suggestion
can be completed as filenames or buffer names, for example.
```

|user-manual|.

------------------------------------------------------------------------------
Some words on the API                                            *lua-guide-api*

The purpose of this guide is to introduce the different ways of interacting
with Neovim through Lua (the "API"). This API consists of three different
layers:

1. The "Vim API" inherited from Vim: |ex-commands| and |builtin-functions| as
well as |user-function|s in Vimscript. These are accessed through |vim.cmd()|
and |vim.fn| respectively, which are discussed under |lua-guide-vimscript|
below.

2. The "Neovim API" written in C for use in remote plugins and GUIs; see |api|.
These functions are accessed through |vim.api|.

3. The "Lua API" written in and specifically for Lua. These are any other
functions accessible through `vim.*` not mentioned already; see |lua-stdlib|.

This distinction is important, as API functions inherit behavior from their
original layer: For example, Neovim API functions always need all arguments to
be specified even if Lua itself allows omitting arguments (which are then
passed as `nil`); and Vim API functions can use 0-based indexing even if Lua

> arrays are 1-indexed by default.

Reviewer 4: This looks a bit confusing. Nvim API uses 0-based indexing much more than Vim API.

Reviewer 2: True; but the contrast here was with Lua, not Nvim API. Do you have a suggestion how to phrase this better?

Reviewer 4: I guess this is fine then.

Through this, any possible interaction can be done through Lua without writing
a complete new API from scratch. For this reason, functions are usually not
duplicated between layers unless there is a significant benefit in
functionality or performance (e.g., you can map Lua functions directly through
|nvim_create_autocmd()| but not through |:autocmd|). In case there are multiple
ways of achieving the same thing, this guide will only cover what is most
convenient to use from Lua.

==============================================================================
Using Lua                                                  *lua-guide-using-Lua*

To run Lua code from the Neovim command line, use the |:lua| command:
>vim
    :lua print("Hello!")
<
Note: each |:lua| command has its own scope and variables declared with the
local keyword are not accessible outside of the command. This won't work:
>vim
    :lua local foo = 1
    :lua print(foo)
    " prints "nil" instead of "1"
<

> You can also use `:lua=`, which is the same as `:lua print(vim.inspect(expr))`,

Reviewer 3: ```suggestion
You can also use `:lua=`, which is the same as `:lua vim.pretty_print(expr)`,
```

Reviewer 2: ...which is the same as `print(vim.inspect(...))` ;)

Reviewer 3: Sure, but I thought it would be nice to mention this function. But it's shown elsewhere in the guide so maybe it's not really necessary here

to conveniently check the value of a variable or a table:
>lua
    :lua=package
<
To run a Lua script in an external file, you can use the |:source| command
exactly like for a Vimscript file:
>vim
    :source ~/programs/baz/myluafile.lua
<

> Finally, you can include Lua code in a vimscript file by putting it inside a

Reviewer 4: ```suggestion
Finally, you can include Lua code in a Vimscript file by putting it inside a
```

      print(data[1])
    end

    vim.fn.jobstart('ls', { on_stdout = print_stdout })
    print(vim.fn.printf('Hello from %s', 'Lua'))
<
This works for both |builtin-functions| and |user-function|s.

Note that hashes (`#`) are not valid characters for identifiers in Lua, so,
e.g., |autoload| functions have to be called with this syntax:
>lua
    vim.fn['my#autoload#function']()
<
------------------------------------------------------------------------------
See also:
• |builtin-functions|: alphabetic list of all Vimscript functions
• |function-list|:     list of all Vimscript functions grouped by topic
• |:runtime|:          run all Lua scripts matching a pattern in |'runtimepath'|
• |package.path|:      list of all paths searched by `require()`

==============================================================================
Variables                                                  *lua-guide-variables*

Variables can be set and read using the following wrappers, which directly
correspond to their |variable-scope|:

• |vim.g|   - global variables (|g:|)
• |vim.b|   - variables for the current buffer (|b:|)
• |vim.w|   - variables for the current window (|w:|)

> • |vim.t|   - variables for the current tabpage(|t:|)

Reviewer 4: ```suggestion
• |vim.t|   - variables for the current tabpage (|t:|)
```

• |vim.v|   - predefined Vim variables (|v:|)
• |vim.env| - environment variables defined in the editor session

Data types are converted automatically. For example:
>lua
    vim.g.some_global_variable = {
      key1 = "value",
      key2 = 300
    }

    print(vim.inspect(vim.g.some_global_variable))
    --> { key1 = "value", key2 = 300 }
<
You can target specific buffers (via number), windows (via |window-ID|), or
tabpages by indexing the wrappers:
>lua
    vim.b[2].myvar = 1            -- set myvar for buffer number 2

>     vim.w[5].myothervar = true    -- set myothervar for window ID 5

Reviewer 4: ```suggestion
    vim.w[1005].myothervar = true    -- set myothervar for window ID 1005
```

>     vim.b[2].myvar = 1            -- set myvar for buffer number 2
>     vim.w[5].myothervar = true    -- set myothervar for window ID 5

Reviewer 4: ```suggestion
    vim.b[2].myvar = 1               -- set myvar for buffer number 2
    vim.w[1005].myothervar = true    -- set myothervar for window ID 1005
```

vim.o

For this reason, there exists a more direct variable-like access using `vim.o`
and friends, similarly to how you can get and set options via `:echo &number`
and `:let &listchars='space:_,tab:>~'`:

• |vim.o|  - behaves like |:set|
• |vim.go| - behaves like |:setglobal|
• |vim.bo| - for buffer-scoped options
• |vim.wo| - for window-scoped options

For example:
>lua
    vim.o.smarttab = false -- :set nosmarttab
    print(vim.o.smarttab)
    --> false
    vim.o.listchars = 'space:_,tab:>~' -- :set listchars='space:_,tab:>~'
    print(vim.o.listchars)
    --> 'space:_,tab:>~'
    vim.o.isfname = vim.o.isfname .. ',@-@' -- :set isfname+=@-@
    print(vim.o.isfname)
    --> '@,48-57,/,.,-,_,+,,,#,$,%,~,=,@-@'
    vim.bo.shiftwidth = 4 -- :setlocal shiftwidth=4
    print(vim.bo.shiftwidth)
    --> 4
<
Just like variables, you can specify a buffer number or |window-ID| for buffer
and window options, respectively. If no number is given, the current buffer or
window is used:
>lua
    vim.bo[4].expandtab = true -- sets expandtab to true in buffer 4
    vim.wo.number = true       -- sets number to true in current window

>     print(vim.wo[0].number)    --> true < Note: |vim.wo| only sets

Author: only sets what?

Reviewer 2: Oops, I seem to have messed up a rebase! (Below as well.)

Reviewer 2: Originally, that was going to be a note about window-scoped local options and a reference to `nvim_set_option_value`. I decided against it in the end.

------------------------------------------------------------------------------
See also:
• |lua-options|

==============================================================================
Defining mappings                                           *lua-guide-mappings*

You can map either Vim commands or Lua functions to key sequences.

------------------------------------------------------------------------------
Creating mappings                                       *lua-guide-mappings-set*

Mappings can be created using |vim.keymap.set()|. This function takes three
mandatory arguments:
• {mode} is a string or a table of strings containing the mode
  prefix for which the mapping will take effect. The prefixes are the ones
  listed in |:map-modes|, or "!" for |:map!|, or empty string for |:map|.
• {lhs} is a string with the key sequences that should trigger the mapping.
  An empty string is equivalent to |<Nop>|, which disables a key.
• {rhs} is either a string with a Vim command or a Lua function that should
  be execucted when the {lhs} is entered.

Examples:
>lua
    -- Normal mode mapping for Vim command

>     vim.keymap.set('n', '<Leader>ex1', 'echo "Example 1"')

Author: I think this needs a `<cmd>` and a `<cr>`

>     -- Normal and Commandline mode mapping for Vim command

Reviewer 4: ```suggestion
    -- Normal and Command-line mode mapping for Vim command
```

>     vim.keymap.set({'n', 'c'}, '<Leader>ex2', 'echo "Example 2"')

Author: Same as above, `<cmd>` and `<cr>`?

    -- Normal mode mapping for Lua function
    vim.keymap.set('n', '<Leader>ex3', vim.treesitter.start)
    -- Normal mode mapping for Lua function with arguments
    vim.keymap.set('n', '<Leader>ex4', function() print('Example 4') end)
<
You can map functions from Lua modules via
>lua
    vim.keymap.set('n', '<Leader>pl1', require('plugin').action)
<
Note that this loads the plugin at the time the mapping is defined. If you
want to defer the loading to the time when the mapping is executed (as for
|autoload| functions), wrap it in `function() end`:
>lua
    vim.keymap.set('n', '<Leader>pl2', function() require('plugin').action() end)
<
The fourth, optional, argument is a table with keys that modify the behavior

> of the mapping such as those from `:map-arguments`. The following are the most

Reviewer 4: ```suggestion
of the mapping such as those from |:map-arguments|. The following are the most
```

useful options:
• `buffer`: If given, only set the mapping for the buffer with the specified
  number; `0` or `true` means the current buffer. >lua
    -- set mapping for the current buffer
    vim.keymap.set('n', '<Leader>pl1', require('plugin').action, { buffer = true})
    -- set mapping for the buffer number 4
    vim.keymap.set('n', '<Leader>pl1', require('plugin').action, { buffer = 4})
<
• `silent`: If set to `true`, suppress output such as error messages. >lua

>     vim.keymap.set('n', '<Leader>pl1', require('plugin').action, { silent = true})

Reviewer 3: ```suggestion
    vim.keymap.set('n', '<Leader>pl1', require('plugin').action, { silent = true })
```

<
• `expr`: If set to `true`, do not execute the {rhs} but use the return value
  as input. Special |keycodes| are converted automatically. For example, the following
  mapping replaces <down> with <c-n> in the popupmenu only: >lua
    vim.keymap.set('c', '<down>', function()
      if vim.fn.pumvisible() == 1 then return '<c-n>' end
      return '<down>'
    end, { expr = true })
<
• `desc`: A string that is shown when listing mappings with, e.g., |:map|.
  This is useful since Lua functions as {rhs} are otherwise only listed as
  `Lua: <number> <source file>:<line>`. Plugins should therefore always use this
  for mappings they create. >lua
    vim.keymap.set('n', '<Leader>pl1', require('plugin').action,
      { desc = 'Execute action from plugin' })
<
• `remap`: By default, all mappings are nonrecursive by default (i.e.,
  |vim.keymap.set()| behaves like |:noremap|). If the {rhs} is itself a mapping
  that should be executed, set `remap = true`: >lua
    vim.keymap.set('n', '<Leader>ex1', '<cmd>echo "Example 1"<cr>')
    -- add a shorter mapping
    vim.keymap.set('n', 'e', '<Leader>ex1', { remap = true })

> <
>   The exception is if the {rhs} is a special |<Plug>| mapping, for which
>   `remap = true` by default: >lua
>     vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)')

Reviewer 4: ```suggestion
```

Reviewer 2: Please explain.

Reviewer 4: There is no such exception

Reviewer 3: `<Plug>` gets expanded even with noremap

Reviewer 4: There is no such exception in the Lua API

Reviewer 2: This is not a discussion of _differences_; it's purely an explanation of how to use the Lua API with as little background as possible. But I can rephrase thist.

Reviewer 2: OK, that makes sense. (It's a bit niche to cover for the purpose of this guide, but it shouldn't be misleading.)

Reviewer 3: Actually one second, I need to confirm this

Reviewer 4: Just like `<SNR>` in `<script>` mappings

<
------------------------------------------------------------------------------
Removing mappings                                       *lua-guide-mappings-del*

A specific mapping can be removed with |vim.keymap.del()|:
>lua
    vim.keymap.del('n', '<Leader>ex1')
    vim.keymap.del({'n', 'c'}, '<Leader>ex2', {buffer = true})
<
------------------------------------------------------------------------------
See also:
• `vim.api.`|nvim_get_keymap()|     - return all global mapping
• `vim.api.`|nvim_buf_get_keymap()| - return all mappings for buffer

==============================================================================
Autocommands                                            *lua-guide-autocommands*

> An |autocommand| is a Vim command or a Lua (or Vimscript) function that is

Author: The two ors feels a bit awkward but unsure what'd be better. Maybe just "Vim command or a Lua/Vimscript function"? I dunno, not that important.

Reviewer 2: Yeah, let's leave Vimscript out in the cold here. Not very relevant in this context.

> executed whenever one or more |events| are triggered automatically in response

Author: Hmm, bit of a wonky phrasing. It makes it sound like the trigger is automatic and not the command/function itself?

Reviewer 2: oops, another rebase pileup. I'll fix!

omitted, it defaults to `pattern = '*'`.
Examples:
>lua
    vim.api.nvim_create_autocmd({"BufEnter", "BufWinEnter"}, {
      pattern = {"*.c", "*.h"},
      command = "echo 'Entering a C or C++ file'",
    })

    -- Same autocommand written with a Lua function instead
    vim.api.nvim_create_autocmd({"BufEnter", "BufWinEnter"}, {
      pattern = {"*.c", "*.h"},
      callback = function() print("Entering a C or C++ file") end,
    })

    -- User event triggered by MyPlugin
    vim.api.nvim_create_autocmd("User", {
      pattern = "MyPlugin",
      callback = function() print("My Plugin Works!") end,
    })
<

Neovim will always call a Lua function with a single table containing information
about the triggered autocommand. The most useful keys are
• `match`: a string that matched the `pattern` (see |<amatch>|)
• `buf`:   the number of the buffer the event was triggered in (see |<abuf>|)
• `file`:  the file name of the buffer the event was triggered in (see |<afile>|)
• `data`:  a table with other relevant data that is passed for some events

For example, this allows you to set buffer-local mappings for some filetypes:
>lua
    vim.api.nvim.create_autocmd("FileType", {

>       pattern = "lua",

Author: nit: hmm, the function is valid as is but should this be `*.lua` instead? Maybe you do want to match only files named `lua`, but perhaps we should use a more recognizable static file name such as `Dockerfile` or `CMakeLists.txt`?

Reviewer 2: It's a `FileType` autocommand, so the pattern _is_ the filetype, not a path glob.

Author: Do you expect me to know everything about vim /s

Creating user commands                           *lua-guide-usercommands-create*

User commands can be created through the Neovim API with
`vim.api.`|nvim_create_user_command()|. This function takes three mandatory
arguments:
• a string that is the name of the command (which must start with an uppercase
  letter to distinguish it from builtin commands);
• a string containing Vim commands or a Lua function that is executed when the
  command is invoked;
• a table with |command-attributes|; in addition, it can contain the keys
  `desc` (a string describing the command); `force` (set to `false` to avoid
  replacing an already existing command with the same name), and `preview` (a
  Lua function that is used for |:command-preview|).

Example:
>lua
    vim.api.nvim_create_user_command('Test', 'echo "It works!"', {})
    vim.cmd.Test()
    --> It works!
<
(Note that the third argument is mandatory even if no attributes are given.)

Lua functions are called with a single table argument containing arguments and
modifiers. The most important are:
• `name`: a string with the command name
• `fargs`: a table containing the command arguments split by whitespace (see |<f-args>|)

> • `bang`: `true` if the command was executed with a `!`  modifier (see |<bang>|)

Author: critical: two whitespaces

detail all intricacies. Think of it as a survival kit - the bare minimum
needed to know to comfortably get started on using Lua in Neovim.

An important thing to note is that this isn't a guide to the Lua language
itself. Rather, this is a guide on how to configure and modify Neovim through
the Lua language and the functions we provide to help with this. Take a look
at |luaref| and |lua-concepts| if you'd like to learn more about Lua itself.
Similarly, this guide assumes some familiarity with the basics of Neovim
(commands, options, mappings, autocommands), which are covered in the
|user-manual|.

------------------------------------------------------------------------------
Some words on the API                                            *lua-guide-api*

The purpose of this guide is to introduce the different ways of interacting
with Neovim through Lua (the "API"). This API consists of three different
layers:

1. The "Vim API" inherited from Vim: |ex-commands| and |builtin| as well as
|user-function|s in Vimscript. These are accessed through |vim.cmd()| and
|vim.fn| respectively, which are discussed under |lua-guide-vimscript| below.

> 2. The "Neovim API" written in C for use in remote plugins and GUIs; see |api|.

Reviewer 3: Is the mention of C needed here?

Reviewer 2: Maybe not needed, but deliberate: It _is_ one of the main differentiators to the Lua API (which is written in Lua), and carries the connotation "fast" (in contrast to the "Vim API" above).

Reviewer 2: (That whole section is still under discussion anyway; it's just a suggestion so far.)

arrays are 1-indexed by default.

Through this, any possible interaction can be done through Lua without writing
a complete new API from scratch. For this reason, functions are usually not
duplicated between layers unless there is a significant benefit in
functionality or performance (e.g., you can map Lua functions directly through
|nvim_create_autocmd| but not through |:autocmd|). In case there are multiple
ways of achieving the same thing, this guide will only cover what is most
convenient to use from Lua.

==============================================================================
Using Lua                                                  *lua-guide-using-Lua*

To run Lua code from the Neovim command line, use the |:lua| command:
>vim
    :lua print("Hello!")
<
Note: each |:lua| command has its own scope and variables declared with the
local keyword are not accessible outside of the command. This won't work:
>vim
    :lua local foo = 1
    :lua print(foo)
    " prints "nil" instead of "1"
<
You can also use `:lua =`, which is the same as `:lua print(vim.inspect(expr))`,
to conveniently check the value of a variable or a table:
>lua

>     :lua =package

Reviewer 3: `:lua=` without the space also works, maybe that should be preferred as it's easier to type?
```suggestion
    :lua=package
```

Reviewer 2: Sure; that's just the way I always type. As long as it's consistent.

• clear (bool): defaults to true. Clear existing commands if group already
                exists.

> Note that an empty table is required even if no options are given for API

Reviewer 2: ```suggestion
Note that an empty table is required even if no options are given for Nvim API
```
